pl_bolts/callbacks/vision/confused_logit.py

```diff
@@ -0,0 +1,129 @@
+
+        For each confused logit $l_i$ the confused images are generated by $\partial l_i/ \partial x$,
+        where $i \in \{0, 1\}$ means the top two confused logits.
+
+        Example::
+
+            from pl_bolts.callbacks.vision import ConfusedLogitCallback
+            trainer = Trainer(callbacks=[ConfusedLogitCallback()])
+
+
+        .. note:: whenever called, this model will look for self.last_batch and self.last_logits in the LightningModule
+
+        .. note:: this callback supports tensorboard only right now
+
+        Args:
+            top_k: How many "offending" images we should plot
+            projection_factor: How much to multiply the input image to make it look more like this logit label
+
+        Authored by:
+
+            - Alfredo Canziani
+
+        """
+        super().__init__()
+        self.top_k = top_k
+        self.projection_factor = projection_factor
+
+    def on_batch_end(self, trainer, pl_module):
+
+        # show images only every 20 batches
+        if (trainer.batch_idx + 1) % 20 != 0:
+            return
+
+        # pick the last batch and logits
+        # TODO: use context instead
+        x, y = pl_module.last_batch
+        l = pl_module.last_logits
+
+        # only check when it has opinions (ie: the logit > 5)
+        if l.max() > 5.0:
+            # pick the top two confused probs
+            (values, idxs) = torch.topk(l, k=2, dim=1)
+
+            # care about only the ones that are at most eps close to each other
+            eps = 0.1
+            mask = (values[:, 0] - values[:, 1]).abs() < eps
+
+            if mask.sum() > 0:
+                # pull out the ones we care about
+                confusing_x = x[mask, ...]
+                confusing_y = y[mask]
+                confusing_l = l[mask]
+
+                mask_idxs = idxs[mask]
+
+                self._plot(confusing_x, confusing_y, trainer, pl_module, mask_idxs)
+
+    def _plot(self, confusing_x, confusing_y, trainer, model, mask_idxs):
+        from matplotlib import pyplot as plt
+
+        batch_size = confusing_x.size(0)
+
+        confusing_x = confusing_x[:self.top_k]
+        confusing_y = confusing_y[:self.top_k]
+
+        model.eval()
+        x_param_a = nn.Parameter(confusing_x)
+        x_param_b = nn.Parameter(confusing_x)
+
+        for logit_i, x_param in enumerate((x_param_a, x_param_b)):
+            l = model(x_param.view(batch_size, -1))
+            l[:, mask_idxs[:, logit_i]].sum().backward()
+
+        # reshape grads
+        grad_a = x_param_a.grad.view(batch_size, 28, 28)
+        grad_b = x_param_b.grad.view(batch_size, 28, 28)
+
+        for img_i in range(len(confusing_x)):
+            x = confusing_x[img_i].squeeze(0)
+            y = confusing_y[img_i]
+            ga = grad_a[img_i]
+            gb = grad_b[img_i]
+
+            mask_idx = mask_idxs[img_i]
+
+            fig = plt.figure(figsize=(15, 10))
```

Comment:
use rather `fig, axarr = plt.subplots(ncol=3, nrow=2)`
and alter you just `axarr[1, 2].imshow(...)`

Reply:
i don't have a lot of time to mess with this. i don't know if this will work, but i know the current version works.
If you feel strongly about it, mind making the changes and posting a colab that shows this works?

Reply:
ok I ll check I because the call to `plt.` takes the last pointer and if you plot two in parallel you will write everything into one